x-pack/plugins/files/public/components/upload_file/upload_state.ts

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  of,
+  map,
+  zip,
+  from,
+  race,
+  take,
+  Subject,
+  finalize,
+  forkJoin,
+  mergeMap,
+  switchMap,
+  catchError,
+  ReplaySubject,
+  BehaviorSubject,
+  type Observable,
+} from 'rxjs';
+import type { FileKind, FileJSON } from '../../../common/types';
+import type { FilesClient } from '../../types';
+import { i18nTexts } from './i18n_texts';
+
+import { createStateSubject, type SimpleStateSubject, parseFileName } from './util';
+
+const prop$ = <T = unknown>(initialValue: T) => new BehaviorSubject<T>(initialValue);
+
+interface FileState {
+  file: File;
+  status: 'idle' | 'uploading' | 'uploaded' | 'upload_failed';
+  id?: string;
+  error?: Error;
+}
+
+export class UploadState {
+  private readonly abort$ = new Subject<void>();
+  private readonly files$$ = prop$<Array<SimpleStateSubject<FileState>>>([]);
```

Comment:
nit: up to you, but a file's state does not necessarily need to be all in one object, could also be something like:

```ts
class Upload {
  public readonly file$ = new BehaviorSubject<File | null>(null);
  public readonly status$ = new BehaviorSubject<'idle' | 'uploading' | 'uploaded' | 'upload_failed'>('idle');
  public readonly id$ = new BehaviorSubject<string>('');
  public readonly error$ = new BehaviorSubject<Error | null>(null);
}
```

and then

```ts
export class UploadState {
  private readonly uploads$ = new BehaviorSubject<Upload[]>([]);
}
```

Reply:
I really like your suggestion! Certainly the types are much cleaner. I think I'll keep the $$ for now, just because I think it maps nicely onto how I have been using files. Happy to change this in future.